scheduler/src/cook/mesos/sandbox.clj

```diff
@@ -182,11 +187,51 @@
   "Asynchronously triggers state syncing from the mesos agent."
-  [{:keys [mesos-agent-query-cache task-id->sandbox-agent]} framework-id agent-hostname]
+  [{:keys [mesos-agent-query-cache pending-sync-agent task-id->sandbox-agent]} framework-id agent-hostname]
   (when (and mesos-agent-query-cache task-id->sandbox-agent framework-id agent-hostname)
     (future
       (refresh-agent-cache-entry
-        framework-id mesos-agent-query-cache task-id->sandbox-agent agent-hostname))))
+        framework-id mesos-agent-query-cache pending-sync-agent task-id->sandbox-agent agent-hostname))))
+
+(defn start-sandbox-publisher
+  "Launches a timer task that triggers publishing of the task-id->sandbox state to datomic.
+   The task is invoked at intervals of publish-interval-ms ms."
+  [task-id->sandbox-agent datomic-conn publish-batch-size publish-interval-ms]
+  (log/info "Starting sandbox publisher at intervals of" publish-interval-ms "ms")
+  (chime/chime-at
+    (periodic/periodic-seq (time/now) (time/millis publish-interval-ms))
+    (fn sandbox-publisher-task [_]
+      (log/info "Requesting publishing of instance sandbox directories")
+      (publish-sandbox-to-datomic! datomic-conn publish-batch-size task-id->sandbox-agent))
+    {:error-handler (fn sandbox-publisher-error-handler [ex]
+                      (log/error ex "Instance sandbox directory publish failed"))}))
+
+(defn start-host-sandbox-syncer
+  "Launches a timer task that triggers syncing of sandbox directories of any pending hosts."
+  [mesos-agent-query-cache pending-sync-agent task-id->sandbox-agent sync-interval-ms max-consecutive-sync-failure]
+  (log/info "Starting sandbox syncer at intervals of" sync-interval-ms "ms")
+  (let [syncer-state {:mesos-agent-query-cache mesos-agent-query-cache
+                      :pending-sync-agent pending-sync-agent
+                      :task-id->sandbox-agent task-id->sandbox-agent}]
+    (chime/chime-at
+      (periodic/periodic-seq (time/now) (time/millis sync-interval-ms))
+      (fn host-sandbox-syncer-task [_]
+        (let [{:keys [framework-id host->consecutive-failures pending-sync-hosts]} @pending-sync-agent]
+          (log/info (count pending-sync-hosts) "hosts have pending syncs")
+          (doseq [agent-hostname pending-sync-hosts]
+            (let [consecutive-failures (get host->consecutive-failures agent-hostname 0)]
+              (cond
+                (>= consecutive-failures max-consecutive-sync-failure)
+                (do
+                  (log/info "Removing entry for" agent-hostname "with" consecutive-failures "consecutive failures")
+                  (send pending-sync-agent clear-pending-sync-hostname agent-hostname :threshold))
+                (not (cache/has? @mesos-agent-query-cache agent-hostname))
+                (do
+                  (log/info "Triggering pending sandbox sync of" agent-hostname)
+                  (send pending-sync-agent clear-pending-sync-hostname agent-hostname :sync)
```

Comment:
Is this call necessary? Should we just let `sync-agent-sandboxes` clean this up?

Reply:
We need this to cleanup pending hosts. I restructured the code a bit to ensure the removal actually happens before the next round of `sync`s are tried.